tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs

```diff
@@ -0,0 +1,142 @@
+
+#nullable enable
+namespace Datadog.Trace.Iast.Telemetry;
+
+internal class ExecutedTelemetryHelper
+{
+    private const string BasicExecutedTag = "_dd.iast.telemetry.";
+    private const string SourceExecutedTag = "executed.source.";
+    private const string SinkExecutedTag = "executed.sink.";
+    private const string PropagationExecutedTag = BasicExecutedTag + "executed.propagation";
+    private static IastMetricsVerbosityLevel _verbosityLevel = Iast.Instance.Settings.IastTelemetryVerbosity;
+    private int[] _executedSinks = new int[Trace.Telemetry.Metrics.IastInstrumentedSinksExtensions.Length];
+    private int[] _executedSources = new int[Trace.Telemetry.Metrics.IastInstrumentedSourcesExtensions.Length];
+    private int _executedPropagations = 0;
+    private object _metricsLock = new();
+
+    public static bool Enabled()
+        => _verbosityLevel <= IastMetricsVerbosityLevel.Information;
+
+    public static bool EnabledDebug()
+        => _verbosityLevel == IastMetricsVerbosityLevel.Debug;
+
+    public void AddExecutedSink(IastInstrumentedSinks type)
+    {
+        if (_verbosityLevel <= IastMetricsVerbosityLevel.Information)
+        {
+            lock (_metricsLock)
+            {
+                _executedSinks[(int)type]++;
+            }
+
+            TelemetryFactory.Metrics.RecordCountIastExecutedSinks(type);
+        }
+    }
+
+    public void AddExecutedPropagation()
+    {
+        if (_verbosityLevel == IastMetricsVerbosityLevel.Debug)
+        {
+            lock (_metricsLock)
+            {
+                _executedPropagations++;
+            }
+
+            TelemetryFactory.Metrics.RecordCountIastExecutedPropagations();
+        }
+    }
+
+    public void AddExecutedSource(IastInstrumentedSources type)
+    {
+        if (_verbosityLevel <= IastMetricsVerbosityLevel.Information)
+        {
+            lock (_metricsLock)
+            {
+                _executedSources[(int)type]++;
+            }
+
+            TelemetryFactory.Metrics.RecordCountIastExecutedSources(type);
+        }
+    }
+
+    public void GenerateMetricTags(List<Tuple<string, int>> tags)
+    {
+        lock (_metricsLock)
+        {
+            if (_executedPropagations > 0)
+            {
+                tags.Add(Tuple.Create(PropagationExecutedTag, _executedPropagations));
+            }
+
+            for (int i = 0; i < _executedSources.Length; i++)
+            {
+                if (_executedSources[i] > 0)
+                {
+                    tags.Add(Tuple.Create(GetExecutedSourceTag((IastInstrumentedSources)i), _executedSources[i]));
+                }
+            }
+
+            for (int i = 0; i < _executedSinks.Length; i++)
+            {
+                if (_executedSinks[i] > 0)
+                {
+                    tags.Add(Tuple.Create(GetExecutedSinkTag((IastInstrumentedSinks)i), _executedSinks[i]));
+                }
+            }
+        }
+
+        ResetMetrics();
+    }
```

Comment:
Can't we add these metrics directly if we pass in the tags?

```suggestion
    public void GenerateMetricTags(ITags tags)
    {
        lock (_metricsLock)
        {
            if (_executedPropagations > 0)
            {
                tags.SetMetric(PropagationExecutedTag, _executedPropagations);
            }

            for (int i = 0; i < _executedSources.Length; i++)
            {
                if (_executedSources[i] > 0)
                {
                    tags.SetMetric(GetExecutedSourceTag((IastInstrumentedSources)i), _executedSources[i]);
                }
            }

            for (int i = 0; i < _executedSinks.Length; i++)
            {
                if (_executedSinks[i] > 0)
                {
                    tags.SetMetric(GetExecutedSinkTag((IastInstrumentedSinks)i), _executedSinks[i]);
                }
            }
        }

        ResetMetrics();
    }
```

Reply:
Ah, ok, I get you know. Yes, I think that it would work. Thanks!